### ISSUE TITLE
DPO

### ISSUE THREAD

**Issue by cemrtkn** [cemrtkn/signifier-trainer#13]
# Overview
Allowing DPO training on top of the current capabilities of the package is targeted now. All capabilities built on top of transformers' trainer library ~~with EM and~~ with token training related features should be enabled. *(EM enablement for DPO is split out to #14 — v1 is plain DPO on an already-merged token checkpoint: the signifier tokens are ordinary vocab rows there, so no wrapper and no phase machinery, and DPO's ~1e-6 lr barely moves them anyway.)*

- What training to run is decided:
```yaml
training:
  model: Qwen/Qwen2.5-7B-Instruct
  ptmp_dir: /ptmp/certuer
  mode: dpo # allowing only dpo and sft defaulting to sft
```
- Its core training logic lives at `finetune/dpo.py`
- It should build on top of the already existing tooling and the [dpo trainer](https://huggingface.co/docs/trl/en/dpo_trainer) as much as possible
- It should allow with  parameter `batch_bidirectionals` packing both directions in one batch so if batch size is 32 then 16 unique question and context pairs and their 2 directions
- It should expose dpo's essential parameters and gate through training parameter `mode` so they should not be set if dpo is not set

## How the DPO knobs ride in `train_args` (no separate dpo block needed)

TRL's `DPOConfig` subclasses `TrainingArguments` — it is TrainingArguments plus `beta`, `loss_type`, `max_prompt_length`, `precompute_ref_log_probs`, etc. So the existing pattern of round-tripping `train_args` through `TrainingArguments` at validation extends directly: in `mode: dpo` the same block round-trips through `DPOConfig` instead. The DPO knobs live inside `train_args`, and in sft mode they are rejected automatically because plain `TrainingArguments` doesn't know them — the "should not be set if dpo is not set" gate comes for free.

## How `batch_bidirectionals` works (same step, not same device)

Both directions of a pair only need to share the same optimizer step, not the same GPU: gradients are all-reduced across ranks before the step, so "in one batch" means the global batch. With the dataset built directions-adjacent (rows 2i and 2i+1 are the two directions of pair i), a sampler that shuffles at pair granularity and emits both members adjacently is sufficient — the DistributedSampler's striding may put the two directions on different ranks, but always in the same step. Requires an even effective batch (per_device × world_size × grad_accum).

## Implementation steps

Small steps, one PR-able chunk each, each building on the last. Step 1 is the dataset builder and lives on the Ekklesia side; new trainer code goes in `finetune/dpo.py`. When `mode` is absent or `sft`, nothing changes anywhere.

- [x] **1. Dataset builder** (Ekklesia, `analysis/scripts/parlacap/training/qa/build_dpo_dataset.py`, new). Turn a Phase-2 cross-generate JSONL into an HF DatasetDict of `prompt` / `chosen` / `rejected` rows, following `build_sft_dataset.py`'s interface as much as necessary (`--qa-runs`-style input, `--output`, `--n-splits`, `--seed`, parquet join for canonical party, `signifier_tokens.txt` sibling, token-signifier prompt rendering). Each cross pair emits its two directions as adjacent rows 2i / 2i+1: prompt carries party A's token (+ shared context and question), chosen = A's answer, rejected = B's answer, then mirrored. The held-out test fold is balanced across party pairings the way `build_cross_eval_set.py` balances — N pairs per unordered pairing, C(5,2) = 10 pairings with BDP excluded as negligible, but the pairing count stays variable, not hard-coded. Fold split at the pair level (both directions always in the same fold), and pairs whose source speeches feed the SFT test fold must not land in DPO train.
- [x] **2. Config gate** (`finetune/sft_types.py`). Add `mode: Literal["sft", "dpo"] = "sft"` and `batch_bidirectionals: bool = False` to `TrainingConfig`. In dpo mode, validate `train_args` as TRL's `DPOConfig` (see above). Guard rails: dpo excludes `em_config.status: true` (#14), `embedding_lr`/`model_lr`, LoRA / quantization / partial fine-tuning — v1 is plain full-FT DPO with a uniform lr; `batch_bidirectionals: true` outside dpo mode is an error. Add a pinned `trl` dependency.
- [x] **3. Dataset path** (`finetune/utils/dataset.py`). In dpo mode the loader expects the step-1 DatasetDict and hands the `prompt` / `chosen` / `rejected` columns straight through — `DPOTrainer` tokenizes with `processing_class` itself, so the SFT parser/collator is bypassed entirely. No new-token resize: the model is a merged token checkpoint whose tokenizer already carries the signifiers, so tokenizer and vocab load as-is.
- [x] **4. Core training logic** (`finetune/dpo.py`, new). `run_dpo` mirroring `run_sft`: load model + tokenizer from the merged checkpoint, build `DPOTrainer` with `ref_model=None` (TRL clones the policy as the frozen reference; `precompute_ref_log_probs: true` is the memory-saving alternative — one scoring pass, then the second model is freed), train, and reuse the barrier-guarded save sequence from `run_sft`. Nothing FSDP-specific beyond what `run_sft` already does — all params train, no wrap-policy changes.
- [x] **5. Bidirectional sampler** (same file). With `batch_bidirectionals: true`, override `_get_train_sampler` with a pair-preserving shuffle: shuffle indices at pair granularity, emit 2i then 2i+1. Assert an even dataset length and an even effective batch.
- [x] **6. Wiring** (`finetune/sft.py`). The entry point dispatches on `config.mode`: `dpo` → `run_dpo`; otherwise the existing `run_sft` path runs exactly as today.
- [x] **7. Tests** (`tests/unit/test_dpo.py`, new). Config guard rails from step 2 (mode gating, DPO knobs rejected in sft mode, `batch_bidirectionals` gating); the pair sampler keeps 2i/2i+1 adjacent under shuffle; the dataset path hands prompt/chosen/rejected through untouched; with `mode` absent, plain SFT still uses the stock path. Builder-side checks (pairing balance, direction adjacency) live with the Ekklesia script.

**Comment by cemrtkn** [comment]
Implemented and merged to main (4970251..a46d750). All 7 steps done; validated end-to-end with a 5% smoke on DAIS (held-out fold-0 reward acc 0.516 -> 0.758 in one epoch). EM-in-DPO follow-up tracked in #14.
